File: app/src/main/java/com/example/security/PasswordAnalysis.kt
package com.example.security

import com.example.database.CredentialPlain
import kotlin.math.ln
import kotlin.math.roundToInt

enum class PasswordIssue {
    EMPTY, REUSED, WEAK
}

data class CredentialFinding(
    val credential: CredentialPlain,
    val issues: List<PasswordIssue>,
    val entropy: Double
)

data class VaultSecurityReport(
    val score: Int,
    val reusedCount: Int,
    val weakCount: Int,
    val emptyCount: Int,
    val totalCount: Int,
    val findings: List<CredentialFinding>
)

object PasswordAnalysis {
    const val WEAK_ENTROPY_BITS = 50.0

    fun estimateEntropyBits(password: String): Double {
        if (password.isEmpty()) return 0.0
        var pool = 0
        if (password.any { it in 'a'..'z' }) pool += 26
        if (password.any { it in 'A'..'Z' }) pool += 26
        if (password.any { it in '0'..'9' }) pool += 10
        if (password.any { !it.isLetterOrDigit() }) pool += 32
        if (pool == 0) return 0.0
        return password.length * (ln(pool.toDouble()) / ln(2.0))
    }

    fun analyzeVault(items: List<CredentialPlain>): VaultSecurityReport {
        if (items.isEmpty()) {
            return VaultSecurityReport(
                score = 100,
                reusedCount = 0,
                weakCount = 0,
                emptyCount = 0,
                totalCount = 0,
                findings = emptyList()
            )
        }

        // Count password frequency for non-empty passwords
        val passwordCounts = mutableMapOf<String, Int>()
        for (item in items) {
            val pass = item.password.trim()
            if (pass.isNotEmpty()) {
                passwordCounts[pass] = (passwordCounts[pass] ?: 0) + 1
            }
        }

        var emptyCount = 0
        var reusedCount = 0
        var weakCount = 0
        val findings = mutableListOf<CredentialFinding>()
        val flaggedItemIds = mutableSetOf<String>()

        for (item in items) {
            val pass = item.password.trim()
            val issues = mutableListOf<PasswordIssue>()
            val entropy = estimateEntropyBits(pass)

            if (pass.isEmpty()) {
                issues.add(PasswordIssue.EMPTY)
                emptyCount++
            } else {
                if ((passwordCounts[pass] ?: 0) > 1) {
                    issues.add(PasswordIssue.REUSED)
                    reusedCount++
                }
                if (entropy < WEAK_ENTROPY_BITS) {
                    issues.add(PasswordIssue.WEAK)
                    weakCount++
                }
            }

            if (issues.isNotEmpty()) {
                flaggedItemIds.add(item.id)
                findings.add(CredentialFinding(credential = item, issues = issues, entropy = entropy))
            }
        }

        // Sort findings: reused first, then empty, then weak; ties broken by lower entropy
        findings.sortWith(
            compareBy<CredentialFinding> { finding ->
                when {
                    finding.issues.contains(PasswordIssue.REUSED) -> 0
                    finding.issues.contains(PasswordIssue.EMPTY) -> 1
                    else -> 2
                }
            }.thenBy { it.entropy }
        )

        val total = items.size
        val flaggedCount = flaggedItemIds.size
        val score = (((total - flaggedCount).toDouble() / total.toDouble()) * 100.0).roundToInt().coerceIn(0, 100)

        return VaultSecurityReport(
            score = score,
            reusedCount = reusedCount,
            weakCount = weakCount,
            emptyCount = emptyCount,
            totalCount = total,
            findings = findings
        )
    }
}

File: src/security/passwordAnalysis.ts
import { CredentialFinding, CredentialPlain, PasswordIssue, VaultSecurityReport } from '../types';

export class PasswordAnalysis {
  static readonly WEAK_ENTROPY_BITS = 50.0;

  static estimateEntropyBits(password: string): number {
    if (!password) return 0.0;
    let pool = 0;
    if (/[a-z]/.test(password)) pool += 26;
    if (/[A-Z]/.test(password)) pool += 26;
    if (/[0-9]/.test(password)) pool += 10;
    if (/[^a-zA-Z0-9]/.test(password)) pool += 32;
    if (pool === 0) return 0.0;
    return password.length * (Math.log(pool) / Math.log(2.0));
  }

  static analyzeVault(items: CredentialPlain[]): VaultSecurityReport {
    // Only analyze login credentials or items with a password field
    const checkableItems = items.filter((item) => item.itemType === 'LOGIN' || item.password);

    if (checkableItems.length === 0) {
      return {
        score: 100,
        reusedCount: 0,
        weakCount: 0,
        emptyCount: 0,
        totalCount: 0,
        findings: [],
      };
    }

    const passwordCounts = new Map<string, number>();
    for (const item of checkableItems) {
      const pass = (item.password || '').trim();
      if (pass.length > 0) {
        passwordCounts.set(pass, (passwordCounts.get(pass) || 0) + 1);
      }
    }

    let emptyCount = 0;
    let reusedCount = 0;
    let weakCount = 0;
    const findings: CredentialFinding[] = [];
    const flaggedItemIds = new Set<string>();

    for (const item of checkableItems) {
      const pass = (item.password || '').trim();
      const issues: PasswordIssue[] = [];
      const entropy = this.estimateEntropyBits(pass);

      if (pass.length === 0) {
        issues.push('EMPTY');
        emptyCount++;
      } else {
        if ((passwordCounts.get(pass) || 0) > 1) {
          issues.push('REUSED');
          reusedCount++;
        }
        if (entropy < this.WEAK_ENTROPY_BITS) {
          issues.push('WEAK');
          weakCount++;
        }
      }

      if (issues.length > 0) {
        flaggedItemIds.add(item.id);
        findings.push({
          credential: item,
          issues,
          entropy,
        });
      }
    }

    // Sort findings: reused first, then empty, then weak; ties broken by lower entropy
    findings.sort((a, b) => {
      const getPriority = (f: CredentialFinding) => {
        if (f.issues.includes('REUSED')) return 0;
        if (f.issues.includes('EMPTY')) return 1;
        return 2;
      };
      const prioDiff = getPriority(a) - getPriority(b);
      if (prioDiff !== 0) return prioDiff;
      return a.entropy - b.entropy;
    });

    const total = checkableItems.length;
    const flaggedCount = flaggedItemIds.size;
    const rawScore = total > 0 ? Math.round(((total - flaggedCount) / total) * 100) : 100;
    const score = Math.max(0, Math.min(100, rawScore));

    return {
      score,
      reusedCount,
      weakCount,
      emptyCount,
      totalCount: total,
      findings,
    };
  }
}
